pep-9999.rst

```diff
@@ -0,0 +1,223 @@
+is generally lost. Nevertheless, there are use cases for keeping records of
+distributions used for installing packages and their provenance.
+
+Python wheels can be built with different compiler flags or supporting
+different wheel tags.  In both cases, users might get into a situation in which
+multiple wheels might be considered by installers (possibly from different
+package indexes) and immediately finding out which wheel file was actually used
+during the installation might be helpful. This way, developers can use
+information about wheels to debug issues making sure the desired wheel
+was actually installed. Another use case could be tools reporting software
+installed, such as tools reporting SBOM (Software Bill of Material), that might
+give more accurate reports.
+
+The motivation described in this PEP is an extension to :pep:`610`.  Besides
+stating information about packages installed using a direct URL, installers SHOULD
+record information also for packages installed from Python package indexes when
+identified by their name, and optionally their version.
+
+Specification
+=============
+
+The ``provenance_url.json`` file SHOULD be created in the ``*.dist-info``
+directory by installers when installing a distribution identified by their
+name, and optionally their version specifier.
+
+This file MUST NOT be created when installing a distribution from a requirement
+specifying a direct URL reference (including a VCS URL).
+
+Only one of ``provenance_url.json`` and ``direct_url.json`` from :pep:`610`
+files MAY be present in ``*.dist-info`` directory.
+
+The ``provenance_url.json`` JSON file MUST be a dictionary, compliant with
+:rfc:`8259` and UTF-8 encoded.
+
+If present, it MUST contain exactly two keys. The first one is ``url``, with
+type ``string``.  The second key MUST be ``archive_info`` with a value defined
+below.
+
+Following :pep:`610`, the ``url`` field MUST be stripped of any sensitive
+authentication information, for security reasons.
+
+The user:password section of the URL MAY however be composed of environment
+variables, matching the following regular expression::
+
+    \$\{[A-Za-z0-9-_]+\}(:\$\{[A-Za-z0-9-_]+\})?
+
+Additionally, the user:password section of the URL MAY be a well-known,
+non-security sensitive string. A typical example is ``git`` in the case of an
+URL such as ``ssh://git@gitlab.com``.
+
+The ``archive_info`` key MUST be present as a dictionary and MUST state at
+least one of the following keys:
+
+- The ``hashes`` key SHOULD be present as a dictionary mapping a hash name to a
+  hex-encoded digest of the file.
+
+  Multiple hashes can be included, and it is up to the consumer to decide what
+  to do with multiple hashes (it may validate all of them or a subset of them, or
+  nothing at all).
+
+  These hash names SHOULD always be normalized to be lowercase.
+
+  Any hash algorithm available via ``hashlib`` (specifically any that can be
```

Comment:
Tighten this up: Lets specify an explicit list of canonical hash name strings that are allowed as what `new()` accepts is too fuzzy as most builds today use OpenSSL to do that lookup which explodes the possibly accepted values into a huge set including mixed case and so on.  That is undesirable.

I suggest only allowing lower case hash method names present as a top level name within the `hashlib` module that accept a single argument.  Also known as hashlib.algorithms_guaranteed minus the multi-argument ones like shake_128 and shake_256 today:

```
>>> sorted(hashlib.algorithms_guaranteed - {'shake_128', 'shake_256'})
['blake2b', 'blake2s', 'md5', 'sha1', 'sha224', 'sha256', 'sha384', 'sha3_224', 'sha3_256', 'sha3_384', 'sha3_512', 'sha512']
>>> for name in sorted(hashlib.algorithms_guaranteed - {'shake_128', 'shake_256'}):
...   getattr(hashlib, name)(b'data').hexdigest()
```

This definition allows new algorithms to be added but excludes the untended non-canonical soup of names that OpenSSL technically allows:

```
>>> hashlib.new('SHA1', b'data').hexdigest()
'a17c9aaa61e80a1bf71d0d850af4e5baa9800bbd'
>>> hashlib.new('ShA1', b'data').hexdigest()
'a17c9aaa61e80a1bf71d0d850af4e5baa9800bbd'
>>> hashlib.new('ShA-1', b'data').hexdigest()
'a17c9aaa61e80a1bf71d0d850af4e5baa9800bbd'
```

none of those are desirable and will complicate things if provided.  we should _require_ our hashlib one true lowercase canonical method name.

Reply:
Tried to address this in [3c9b67f](https://github.com/fridex/peps/pull/2/commits/3c9b67f50f1c3d56be53a5c89bd7cf1c8a2f836d). [hashlib docs state explicitly canonical names](https://docs.python.org/3/library/hashlib.html#hashlib.hash.name) so assuming the current wording is okay.

Resolving this discussion, feel free to reopen if additional work is needed. Thank you!